### ISSUE TITLE
http/2 websocket

### ISSUE THREAD

**Issue by dottodot** [flynn/flynn#3053]
So i've just updated to the latest version that supports http/2 

Would this break web sockets as I'm not getting an error of

> failed: WebSocket opening handshake was canceled

and it's the only thing I can think of that would cause this.

**Comment by titanous** [comment]
I assume that error is coming from the client? Can you confirm from the network debug information in the browser what protocol was used? My understanding is that Websockets do not work over HTTP/2, and I would expect browsers not to try that, but I'm not certain.

**Comment by dottodot** [comment]
I'm using socket.io with the websocket transport on the client.

Is it possible to disable http/2 until I can work it out?

**Comment by titanous** [comment]
Can you check the router log for any errors? There is not currently a flag, but I can add one today (ideally I'd like to understand the issue too). I'll do some local testing, what browser are you using?

**Comment by dottodot** [comment]
I'm using chrome and just switched back to my old cluster for the time being, but I'll have a look at the logs in the next hour or so.

**Comment by dottodot** [comment]
Ok this is the router log from 1 node on my cluster 

```
root@cluster7:~# 2016/07/05 16:31:08 http: TLS handshake error from 88.109.148.2:62280: EOF
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:34:31 http: TLS handshake error from 173.46.225.169:57390: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:34:31 http: TLS handshake error from 173.46.225.169:57392: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:34:40 http2: server: error reading preface from client 173.46.225.169:57399: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:34:40 http2: server: error reading preface from client 173.46.225.169:57400: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:34:40 http: TLS handshake error from 173.46.225.169:57401: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:35:13 http2: server: error reading preface from client 173.46.225.169:57429: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:35:13 http2: server: error reading preface from client 173.46.225.169:57430: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:35:13 http2: server: error reading preface from client 173.46.225.169:57428: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:35:13 http: TLS handshake error from 173.46.225.169:57432: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:35:13 http2: server: error reading preface from client 173.46.225.169:57431: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:35:26 http: TLS handshake error from 173.46.225.169:57445: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:35:26 http: TLS handshake error from 173.46.225.169:57446: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:35:26 http: TLS handshake error from 173.46.225.169:57448: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:35:26 http: TLS handshake error from 173.46.225.169:57451: remote error: unknown certificate authority
-bash: 2016/07/05: No such file or directory
root@cluster7:~# 2016/07/05 16:35:31 http: TLS handshake error from 77.102.159.122:52693: EOF
-bash: 2016/07/05: No such file or directory
root@cluster7:~# t=2016-07-05T16:39:45+0000 lvl=eror msg="heartbeat failed" component=discoverd service=router-http addr=82.223.37.254:80 err="unknown_error: Something went wrong"
root@cluster7:~# t=2016-07-05T16:39:45+0000 lvl=eror msg="heartbeat failed" component=discoverd service=router-http addr=82.223.37.254:80 err="unknown_error: Something went wrong"
root@cluster7:~# t=2016-07-05T16:39:46+0000 lvl=eror msg="heartbeat failed" component=discoverd service=router-http addr=82.223.37.254:80 err="unknown_error: Something went wrong"
```
so there's something not right by the looks of it but then I've just found that I'm seeing a lot of TLS handshake errors on my current cluster which I didn't realise.

**Comment by titanous** [comment]
You can ignore the handshake errors, those are probably just scanners or other bots that are hitting weird routes. I've got socket.io running on a local cluster and I'm testing it out now.

**Comment by titanous** [comment]
I've tested the latest socket.io in Firefox and Chrome with the new HTTP/2 code, and the browsers appear to be working properly. Two sockets are created, one for HTTP/2 and one for the HTTP/1.1 websocket upgrade, all communications in the socket.io chat example are going over websockets in both browsers.

Do you have a live example that I can look at? Are you doing any specific configuration of socket.io on the server or client?

**Comment by dottodot** [comment]
I'm using sailsjs so could be something to do with that as it's not a standard socket.io setup, or I'm wondering if it's an ssl issue as I've read that any kind of mismatch would cause it to fail. I have a letencrypt/certbot certificate installed but there didn't seem to be any issue with that when I moved to the new cluster.

The site is https://www.baileyofsheffield.com but it's currently on the old cluster.

**Comment by titanous** [comment]
There is a known issue with restoring from backups in the most recent version that can cause the TLS certificate/key to be lost, I'd suggest running `flynn route update` with the cert/key that you want for the route and trying again. This issue will be fixed by #3049 today.

**Comment by dottodot** [comment]
Yep that was it all working now.
